packages/test/test-utils/src/opProcessingController.ts

```diff
@@ -10,6 +10,49 @@ import { ILocalDeltaConnectionServer } from "@fluidframework/server-local-server
 // An IDeltaManager alias to be used within this class.
 export type DeltaManager = IDeltaManager<ISequencedDocumentMessage, IDocumentMessage>;
 
+class DeltaManagerToggle {
+    private inboundPauseP: Promise<void> | undefined;
+    private outboundPauseP: Promise<void> | undefined;
+    constructor(public readonly deltaManager: DeltaManager) {
+    }
+
+    public async togglePauseAll() {
+        return Promise.all([this.togglePauseInbound(), this.togglePauseOutbound()]);
+    }
+
+    public toggleResumeAll() {
+        this.toggleResumeInbound();
+        this.toggleResumeOutbound();
+    }
+    public async togglePauseInbound() {
+        if (!this.inboundPauseP) {
+            this.inboundPauseP = this.deltaManager.inbound.pause();
+        }
+        return this.inboundPauseP;
+    }
+
+    public async togglePauseOutbound() {
+        if (!this.outboundPauseP) {
+            this.outboundPauseP = this.deltaManager.outbound.pause();
+        }
+        return this.outboundPauseP;
+    }
+
+    public toggleResumeInbound() {
+        if (this.inboundPauseP) {
+            this.inboundPauseP = undefined;
```

Comment:
This implementation assumes that ref-count changes happen synchronously when calling pause() , i.e. ref-counted increase is reflected right away, and thus it's safe to call resume before promise is satisfied.
I feel like all users of pause() should always wait for it to complete before calling resume(). If so, it would be great to assert that requirement here somehow.

Reply:
I am not sure that is always the case.  e.g. (DeltaScheduler.batchEnd)[https://github.com/microsoft/FluidFramework/blob/7f356074c560f83803710ea3320aa9bd41aa3b63/packages/runtime/container-runtime/src/deltaScheduler.ts#L88]